command/build.go

```diff
@@ -186,6 +186,11 @@ func (c *BuildCommand) RunContext(buildCtx context.Context, cla *BuildArgs) int
 	log.Printf("Force build: %v", cla.Force)
 	log.Printf("On error: %v", cla.OnError)
 
+	if len(builds) == 0 {
+		c.Ui.Say("Warning: no builds to run, exiting.")
+		return ret
+	}
+
```

Comment:
nit: The warning message while good doesn't provide much context to a new user. Maybe we can add something like the following.
```suggestion
		if len(builds) == 0 {
		diags := hcl.Diagnostics{
			&hcl.Diagnostic{
				Summary:  "Missing build block",
				Detail:   "A build block with one or more sources is required for executing a build.",
				Severity: hcl.DiagError,
			},
		}
		return writeDiags(c.Ui, nil, diags)
	}

```

FYI I like to use diagnostics for these types of errors as they are setup to provide additional context without having to build long error messages.

Reply:
Unfortunately having this error does not necessarily mean that there's not a build block defined, but rather that no builds are to be run, this also happens when running a command with only/except that yields no builds to run.
I do agree that making the warning an error makes sense, I'll fold that in with your suggestion to replace the ui.Say by a hcl.Diagnostic, but I'll amend the error message at the same time.

Reply:
Come to think of it, we can probably add one more error message for the no build block case, we can get that information from the PackerConfig structure, I'll add this and reroll this PR

Reply:
Ah got it! That is a good call out. I remember now you mentioning the only/except case.